Basic dynamo support for traceable collectives

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94440

Make traceable collectives work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
For now, wait happens immediately, which generally forces an early sync.

Later, find a way either in dynamo or AOT stack to handle
AsyncCollectiveTensor to get the wait in the optimal place.

Note on implementation:
- Dynamo traces 'user-level' fc apis that are designed to behave differently
  in eager vs compiled.  In eager, there will be work-obj registration and
  a wrapper subclass will insert a 'wait' call at the appropriate time.
  In compile/trace mode, wait will be immetiately called, and work obj
  registration is required to be handled by the compile backend at runtime.
- Dynamo needs to trace into some of the helper functions in the 'user-level'
  api, such as '_expand_group' which is essentially a constant transformation.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

wconstab committed:
Basic dynamo support for traceable allreduce

Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable allreduce"


Attempt to make traceable allreduce work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
The long term solution should be "torchdispatch tensor subclass support", which is WIP

Option 1: Dynamo inlines Traceable Collectives, and handles AsyncTensor subclass
- AsyncTensor needs special handling since it is an unsupported subclass
- Special-case Dynamo to perform the simple function of AsyncTensor on its behalf:
  insert a 'Wait' op either "right after" the op that returned AsyncTensor (least optimal)
  or right before first user of AsyncTensor (optimal but more complicated)

  (a) add 'torch.distributed.traceable_collectives'
	  to `_dynamo.config.skipfiles_inline_module_allowlist`
  (b) implement AsyncTensor special behavior

Option 2: Dynamo allows traceable collectives in graph, doesn't inline
- we avoid Dynamo worrying about AsyncTensor at all. Just let it come up later,
  when doing dispatch tracing during AotAutograd
  (a) maybe the default already allows this op in graph!
  (b) still need AsyncTensor handling? Bc we seem to be faketensor tracing the op,
	  which fails when wrapping a FakeTensor in an AsyncTensor

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable collectives"


Make traceable collectives work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
For now, wait happens immediately, which generally forces an early sync.

Later, find a way either in dynamo or AOT stack to handle
AsyncCollectiveTensor to get the wait in the optimal place.

Note on implementation:
- Dynamo traces 'user-level' fc apis that are designed to behave differently
  in eager vs compiled.  In eager, there will be work-obj registration and
  a wrapper subclass will insert a 'wait' call at the appropriate time.
  In compile/trace mode, wait will be immetiately called, and work obj
  registration is required to be handled by the compile backend at runtime.
- Dynamo needs to trace into some of the helper functions in the 'user-level'
  api, such as '_expand_group' which is essentially a constant transformation.

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable collectives"


Make traceable collectives work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
For now, wait happens immediately, which generally forces an early sync.

Later, find a way either in dynamo or AOT stack to handle
AsyncCollectiveTensor to get the wait in the optimal place.

Note on implementation:
- Dynamo traces 'user-level' fc apis that are designed to behave differently
  in eager vs compiled.  In eager, there will be work-obj registration and
  a wrapper subclass will insert a 'wait' call at the appropriate time.
  In compile/trace mode, wait will be immetiately called, and work obj
  registration is required to be handled by the compile backend at runtime.
- Dynamo needs to trace into some of the helper functions in the 'user-level'
  api, such as '_expand_group' which is essentially a constant transformation.

[ghstack-poisoned]

kumpera commented:
Maybe there's an alternative to having dynamo fully handling AsyncCollectiveTensor.

We could:
1) Do wait scheduling in inductor's on its FX graph.
2) If a tensor is not used in the FX graph, escape it as an AsyncCollectiveTensor

wconstab commented:
> Maybe there's an alternative to having dynamo fully handling AsyncCollectiveTensor.

I'm not sure which version of this PR you read, or if my update failed to push. My earlier approach was to handle AsyncCollectiveTensor in dynamo, but for the latest PR it is just skipped and dynamo traces the collective and the wait immediately.

wconstab committed:
Update on "Basic dynamo support for traceable collectives"


Make traceable collectives work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
For now, wait happens immediately, which generally forces an early sync.

Later, find a way either in dynamo or AOT stack to handle
AsyncCollectiveTensor to get the wait in the optimal place.

Note on implementation:
- Dynamo traces 'user-level' fc apis that are designed to behave differently
  in eager vs compiled.  In eager, there will be work-obj registration and
  a wrapper subclass will insert a 'wait' call at the appropriate time.
  In compile/trace mode, wait will be immetiately called, and work obj
  registration is required to be handled by the compile backend at runtime.
- Dynamo needs to trace into some of the helper functions in the 'user-level'
  api, such as '_expand_group' which is essentially a constant transformation.

[ghstack-poisoned]

wconstab commented on torch/_dynamo/allowed_functions.py:
```diff
@@ -253,6 +254,16 @@ def is_allowed(obj):
     # torch.ops is populated lazily so we don't necessarily have them in
     # _allowed_function_ids.  Figure it out by testing the type instead
     # in those cases
+    if obj in (
```
@anijain2305 I didn't want to have to modify `is_allowed` function at all, so I didn't worry about doing the modification in a clean/maintainable way.  I want to find a way to inline all these functions.  What's the best way to do this?

I think it's cleaner from readability perspective if I apply a decorator inside _functional_collectives.py on each of these functions, as that will enable anyone adding new functions to discover/copy/paste and this way is easy to miss.

anijain2305 replied:
This diff works 

~~~
diff --git a/torch/_dynamo/allowed_functions.py b/torch/_dynamo/allowed_functions.py
index d81a2ab0f87..4fe693a64f1 100644
--- a/torch/_dynamo/allowed_functions.py
+++ b/torch/_dynamo/allowed_functions.py
@@ -109,22 +109,9 @@ def _disallowed_function_ids():
         torch._C._dynamo.eval_frame.unsupported,
     ]

-    try:
+    if torch.distributed.is_available():
         from torch.distributed import _functional_collectives
-
-        remove.extend(
-            [
-                _functional_collectives.all_reduce,
-                _functional_collectives.all_gather_tensor,
-                _functional_collectives.reduce_scatter_tensor,
-                _functional_collectives._expand_group,
-                _functional_collectives._maybe_wrap_tensor,
-                _functional_collectives._are_we_tracing,
-            ]
-        )
-        config.skipfiles_inline_module_allowlist.append(_functional_collectives)
-    except ImportError:
-        pass
+        config.skipfiles_inline_module_allowlist.add(_functional_collectives)

     # extract all dtypes from torch
     dtypes = [
@@ -164,7 +151,6 @@ def _allowed_function_ids():
             "torch._C.inductor.",
             "torch.fx.",
             "torch.distributed.fsdp.",
-            "torch.distributed._functional_collectives.",
         )
         allowed_modules_dot = tuple([x + "." for x in allowed_modules])
         module = inspect.getmodule(obj)
~~~

The generated Dynamo graphs look like these

~~~
 ===== __compiled_fn_2 =====
 <eval_with_key>.5 class GraphModule(torch.nn.Module):
    def forward(self, L_inp_ : torch.Tensor):
        l_inp_ = L_inp_

        # File: /scratch/anijain/work/pytorch/torch/distributed/_functional_collectives.py:432, code: tensor = torch.ops.c10d_functional.reduce_scatter_tensor(self, reduceOp, tag, rankset, group_size)  # type: ignore[attr-defined]
        reduce_scatter_tensor = torch.ops.c10d_functional.reduce_scatter_tensor(l_inp_, 'sum', '', [0], 1);  l_inp_ = None

        # File: /scratch/anijain/work/pytorch/torch/distributed/_functional_collectives.py:345, code: return torch.ops.c10d_functional.wait_tensor(tensor)  # type: ignore[attr-defined]
        wait_tensor = torch.ops.c10d_functional.wait_tensor(reduce_scatter_tensor);  reduce_scatter_tensor = None
        return (wait_tensor,)
        
~~~

wconstab replied:
Yep, that works. Thanks @anijain2305!

wconstab committed:
Update on "Basic dynamo support for traceable collectives"


Make traceable collectives work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
For now, wait happens immediately, which generally forces an early sync.

Later, find a way either in dynamo or AOT stack to handle
AsyncCollectiveTensor to get the wait in the optimal place.

Note on implementation:
- Dynamo traces 'user-level' fc apis that are designed to behave differently
  in eager vs compiled.  In eager, there will be work-obj registration and
  a wrapper subclass will insert a 'wait' call at the appropriate time.
  In compile/trace mode, wait will be immetiately called, and work obj
  registration is required to be handled by the compile backend at runtime.
- Dynamo needs to trace into some of the helper functions in the 'user-level'
  api, such as '_expand_group' which is essentially a constant transformation.

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable collectives"


Make traceable collectives work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
For now, wait happens immediately, which generally forces an early sync.

Later, find a way either in dynamo or AOT stack to handle
AsyncCollectiveTensor to get the wait in the optimal place.

Note on implementation:
- Dynamo traces 'user-level' fc apis that are designed to behave differently
  in eager vs compiled.  In eager, there will be work-obj registration and
  a wrapper subclass will insert a 'wait' call at the appropriate time.
  In compile/trace mode, wait will be immetiately called, and work obj
  registration is required to be handled by the compile backend at runtime.
- Dynamo needs to trace into some of the helper functions in the 'user-level'
  api, such as '_expand_group' which is essentially a constant transformation.

[ghstack-poisoned]

kumpera reviewed (approved):
I'm not entirely sure on wh

wconstab committed:
Update on "Basic dynamo support for traceable collectives"


Make traceable collectives work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
For now, wait happens immediately, which generally forces an early sync.

Later, find a way either in dynamo or AOT stack to handle
AsyncCollectiveTensor to get the wait in the optimal place.

Note on implementation:
- Dynamo traces 'user-level' fc apis that are designed to behave differently
  in eager vs compiled.  In eager, there will be work-obj registration and
  a wrapper subclass will insert a 'wait' call at the appropriate time.
  In compile/trace mode, wait will be immetiately called, and work obj
  registration is required to be handled by the compile backend at runtime.
- Dynamo needs to trace into some of the helper functions in the 'user-level'
  api, such as '_expand_group' which is essentially a constant transformation.

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable collectives"


Make traceable collectives work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
For now, wait happens immediately, which generally forces an early sync.

Later, find a way either in dynamo or AOT stack to handle
AsyncCollectiveTensor to get the wait in the optimal place.

Note on implementation:
- Dynamo traces 'user-level' fc apis that are designed to behave differently
  in eager vs compiled.  In eager, there will be work-obj registration and
  a wrapper subclass will insert a 'wait' call at the appropriate time.
  In compile/trace mode, wait will be immetiately called, and work obj
  registration is required to be handled by the compile backend at runtime.
- Dynamo needs to trace into some of the helper functions in the 'user-level'
  api, such as '_expand_group' which is essentially a constant transformation.

[ghstack-poisoned]

wconstab committed:
Update on "Basic dynamo support for traceable collectives"


Make traceable collectives work with torchdynamo,
bypassing problems with tracing the AsyncTensor subclass.

Accept a suboptimal solution for now, and optimize it later.
For now, wait happens immediately, which generally forces an early sync.

Later, find a way either in dynamo or AOT stack to handle
AsyncCollectiveTensor to get the wait in the optimal place.

Note on implementation:
- Dynamo traces 'user-level' fc apis that are designed to behave differently
  in eager vs compiled.  In eager, there will be work-obj registration and
  a wrapper subclass will insert a 'wait' call at the appropriate time.
  In compile/trace mode, wait will be immetiately called, and work obj
  registration is required to be handled by the compile backend at runtime.
- Dynamo needs to trace into some of the helper functions in the 'user-level'
  api, such as '_expand_group' which is essentially a constant transformation.

[ghstack-poisoned]

wconstab commented on torch/_dynamo/allowed_functions.py:
```diff
@@ -108,6 +108,23 @@ def _disallowed_function_ids():
     ]
+
+    try:
+        from torch.distributed import _functional_collectives
```
fixed. thanks!

wconstab commented:
@pytorchbot merge
